File: ordenamiento/c#/mergesort.cs
using System;
class MergeSort
{
    // merge dos subarrays de a[]
    static void Merge(int[] a, int l, int m, int r)
    {
        int a1 = m - l + 1; // tamaño del primer subarray
        int a2 = r - m; // tamaño del segundo subarray

        // crear arrays temporales
        int[] L = new int[a1]; // crea el array temporal
        int[] R = new int[a2]; // crea el array temporal

        // copiar datos a los arrays temporales L[] y R[]
        for (int j = 0; j < a1; j++)
            L[j] = a[l + j];
        for (int k = 0; k < a2; k++) // copiar datos al array temporal
            R[k] = a[m + 1 + k];

        int i = 0; // indice inicial del primer subarray
        int jj = 0; // indice inicial del segundo subarray
        int kk = l; // indice inicial del subarray mezclado

        // mezclar los arrays temporales de nuevo en a[l..r]
        while (i < a1 && jj < a2) // recorrer ambos arrays
        {
            if (L[i] <= R[jj]) // comparar los elementos de ambos arrays
            {
                a[kk] = L[i]; // copiar el elemento mas pequeño al array original
                i = i + 1; // incrementar el indice del primer array
            }
            else // si el elemento del segundo array es mas pequeño
            {
                a[kk] = R[jj]; // copiar el elemento mas pequeño del array original
                jj = jj + 1; // incrementar el indice del segundo array
            }
            kk = kk + 1; // incrementar el indice del array original
        }
        // copiar los elementos restantes de L[], si hay alguno
        while (i < a1) // copiar los elementos restantes del primer array
        {
            a[kk] = L[i]; // copiar el elemento al array original
            i = i + 1; // incrementar el indice del primer array
            kk = kk + 1; // incrementar el indice del array original
        }
    }

    // l es para el indice izquierdo y r es para el indice derecho del subarray de 'a' a ser ordenado
    static void MergeSortMethod(int[] a, int l, int r) // funcion principal que ordena a[l..r]
    {
        if (l < r)
        {
            // igual que (l + r)/2, pero evita el desbordamiento para grandes valores de l y h
            int m = l + (r - l) / 2;
            // ordenar la primera y segunda mitad
            MergeSortMethod(a, l, m); // ordenar la primera mitad
            MergeSortMethod(a, m + 1, r); // ordenar la segunda mitad
            Merge(a, l, m, r); // mezclar las dos mitades
        }
    }

    static void Main(string[] args)
    {
        int[] a = {39, 28, 44, 11};
        int s = a.Length;

        Console.WriteLine("Antes de ordenar el arreglo: ");
        for (int j = 0; j < s; j++)
            Console.Write(a[j] + " ");

        MergeSortMethod(a, 0, s - 1);

        Console.WriteLine("\nDespués de ordenar el arreglo: ");
        for (int j = 0; j < s; j++)
            Console.Write(a[j] + " ");
    
    }
}

File: mergesort.js
//merge dos subarrays de a[]
function merge (a, l, m, r)
{
    const a1 = m - l + 1;
    const a2 = r - m;

    //crear arrays temporales
    const L = new Array(a1).fill(0);
    const R = new Array(a2).fill(0);

    // copiar datos a los arrays temporales L[] y R[]
    for (let j = 0; j < a1; j++)
        L[j] = a[l + j];
    for (let k = 0; k < a2; k++) // copiar datos al array temporal
        R[k] = a[m + 1 + k];

    let i = 0; // indice inicial del primer subarray
    let j = 0; // indice inicial del segundo subarray
    let k = l; // indice inicial del subarray mezclado

    // mezclar los arrays temporales de nuevo en a[l..r]
    while (i < a1 && j < a2) { // recorrer ambos arrays
        if (L[i] <= R[j]) { // comparar los elementos de ambos arrays
            a[k] = L[i]; // copiar el elemento mas pequeño al array original
            i = i + 1; // incrementar el indice del primer array
        } else { // si el elemento del segundo array es mas pequeño
            a[k] = R[j]; // copiar el elemento mas pequeño del array original
            j = j + 1; // incrementar el indice del segundo array
        }
        k = k + 1; // incrementar el indice del array original
    }
    // copiar los elementos restantes de L[], si hay alguno
    while (i < a1) { // copiar los elementos restantes del primer array
        a[k] = L[i]; // copiar el elemento al array original
        i = i + 1; // incrementar el indice del primer array
        k = k + 1; // incrementar el indice del array original
    }
}

// l es para el indice izquierdo y r es para el indice derecho del subarray de 'a' a ser ordenado
function mergeSort(a, l, r) // funcion principal que ordena a[l..r]
{ 
    if (l < r) {
        // igual que (l + r)/2, pero evita el desbordamiento para grandes valores de l y h
        const m = l + Math.floor((r - l) / 2);
        // ordenar la primera y segunda mitad
        mergeSort(a, l, m); // ordenar la primera mitad
        mergeSort(a, m + 1, r); // ordenar la segunda mitad
        merge(a, l, m, r); // mezclar las dos mitades
    }
}

//codigo para probar la implementacion de mergesort
const a = [39, 28, 44, 11];
const s = a.length;

console.log("Antes de ordenar el arreglo: ");
console.log(a.join(" "));

mergeSort(a, 0, s - 1);

console.log("Despues de ordenar el arreglo: ");
console.log(a.join(" "));